### PR TITLE
Introduce Insights and Personalization feature

### PR DESCRIPTION
**Pull request by julienbourdeau**
Introduce Personalization and Insights client

### Personalization

```ruby

client.set_strategy({...})
strategy = client.get_strategy()
```

### Insights

```ruby
@insights = Algolia::Insights.new({
                                     :application_id => ENV['ALGOLIA_APP_ID']
                                     :api_key => ENV['ALGOLIA_API_KEY']
                                 })
user_insights = @insights.user(user_token)

# Click events (you can pass request options to every call
user_insights.clicked_filters("eventName", "indexName", filters);
user_insights.clicked_object_ids("eventName", "indexName", objectIDs);
user_insights.clicked_object_ids_after_search("eventName", "indexName", objectIDs, positions, "queryID");

# Conversion events
user_insights.converted_filters("eventName", "indexName", filters);
user_insights.converted_object_ids("eventName", "indexName", objectIDs);
user_insights.converted_object_idsafter_search("eventName", "indexName", objectIDs, "queryID");

# View events
user_insights.viewed_filters("eventName", "indexName", filters);
user_insights.viewed_object_ids("eventName", "indexName", objectIDs);
```

### CODE REVIEW

File: lib/algolia/insights.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                                })
+    end
+
+    def user(user_token)
+      UserInsights.new(self, user_token)
+    end
+
+    def send_event(event)
+      puts event.inspect
+      exit(1)
```
Comment (nunomaduro) [inline lib/algolia/insights.rb]: I believe this is test code right? It should be removed/

File: lib/algolia/insights.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      send_vents([event])
```
Comment (nunomaduro) [inline lib/algolia/insights.rb]: This should be `send_events`.

File: lib/algolia/insights.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    end
+
+    def send_events(events)
```
Comment (nunomaduro) [inline lib/algolia/insights.rb]: No `request_options = {}`?

Comment (nunomaduro) [reply lib/algolia/insights.rb]: This comment goes for all methods that should have `request_options = {}`?

File: lib/algolia/insights.rb [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+      if response.code / 100 != 2
+        raise AlgoliaProtocolError.new(response.code, "Cannot #{method} to #{url}: #{response.content}")
+      end
+
+      JSON.parse(response.content)
+    end
+  end
+
+  class UserInsights
```
Comment (nunomaduro) [inline lib/algolia/insights.rb]: It's idiomatic in Ruby have 2 classes in the same file?

File: lib/algolia/insights.rb [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                  'positions' => positions,
+                  'queryId' => query_id,
+              }, event_name, index_name)
+    end
+
+    def clicked_filters(event_name, index_name, filters)
+      clicked({ 'filters' => filters }, event_name, index_name)
+    end
+
+    def converted_object_ids(event_name, index_name, object_ids)
```
Comment (nunomaduro) [inline lib/algolia/insights.rb]: Missing method `converted_object_ids_after_search`.

File: lib/algolia/insights.rb [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    def viewed(event, event_name, index_name)
+      send_vent(event.merge({
+                                'eventType' => 'view',
+                                'eventName' => event_name,
+                                'index' => index_name,
+                            }))
+    end
+
+    def send_vent(event)
```
Comment (nunomaduro) [inline lib/algolia/insights.rb]: This should be `send_event`.

File: lib/algolia/client.rb [hunk 7]
```diff
@@ -491,9 +491,17 @@ def wait_task(index_name, taskID, time_before_retry = WAIT_TASK_DEFAULT_TIME_BEF
       end
```
Comment (nunomaduro) [inline lib/algolia/client.rb]: You may want to add tests on insights, just to make sure that methods signatures exist.

File: lib/algolia/insights.rb (hunk 1 above)
Comment (julienbourdeau) [reply lib/algolia/insights.rb]: 🤣

File: lib/algolia/client.rb (hunk 7 above)
Comment (julienbourdeau) [reply lib/algolia/client.rb]: I'm not sure how to test that

**Comment by julienbourdeau** [comment]
Thank you @nunomaduro for your review and sorry for the sloppy PR 👍

File: lib/algolia/client.rb (hunk 7 above)
Comment (nunomaduro) [reply lib/algolia/client.rb]: @julienbourdeau Just call the method, that is good just to make sure that mistakes as renaming `send_vent` are catched.

**Comment by nunomaduro** [comment]
@julienbourdeau Thanks buddy, just saw that you have resolved all feedback. I think we should consider my feedback about testing - just to make sure that mistakes as renaming send_vent are catched.

Example: https://github.com/algolia/algoliasearch-client-python/blob/500f17de2db43b6a36019a74ad13c767d9c33c41/tests/test_insights_client.py.

**Comment by julienbourdeau** [comment]
https://github.com/algolia/algoliasearch-client-ruby/pull/326/commits/e7aada612553faa01b2369fa6fa25ddda81927b8

**Comment by julienbourdeau** [comment]
I have no idea what's going on with ruby 1.9 on travis, but this is not related to this PR, and it works for 1.8 so I'm going to merge anyway.

File: lib/algolia/insights.rb [hunk 8]
```diff
@@ -0,0 +1,131 @@
+module Algolia
+
+  class Insights
+    MIN_RUBY_VERSION = '1.9.0'
```
Comment (raphi) [inline lib/algolia/insights.rb]: Why are we setting a hard minimum version?

File: lib/algolia/insights.rb (hunk 4 above)
Comment (raphi) [reply lib/algolia/insights.rb]: Actually it's not a good practice in Ruby (algolia rails isn't a good reference^^)

**Pull request merged**